File: ch10/py/rag_graph.py
from typing import List, TypedDict
from langchain_community.document_loaders import WebBaseLoader
from langchain.schema import Document
from langgraph.graph import END, StateGraph, START
from langchain_community.vectorstores import InMemoryVectorStore
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_openai import OpenAIEmbeddings
from langchain import hub
from langchain_openai import ChatOpenAI


class GraphState(TypedDict):
    """
    Represents the state of our graph.

    Attributes:
        question: question
        scraped_documents: list of documents
        vectorstore: vectorstore
    """

    question: str
    scraped_documents: List[str]
    vectorstore: InMemoryVectorStore
    answer: str


def scrape_blog_posts(state) -> List[Document]:
    """
    Scrape the blog posts and create a list of documents
    """

    urls = [
        "https://blog.langchain.dev/top-5-langgraph-agents-in-production-2024/",
        "https://blog.langchain.dev/langchain-state-of-ai-2024/",
        "https://blog.langchain.dev/introducing-ambient-agents/",
    ]

    docs = [WebBaseLoader(url).load() for url in urls]
    docs_list = [item for sublist in docs for item in sublist]

    return {"scraped_documents": docs_list}


def indexing(state):
    """
    Index the documents
    """
    text_splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
        chunk_size=250, chunk_overlap=0
    )
    doc_splits = text_splitter.split_documents(state["scraped_documents"])

# Add to vectorDB
    vectorstore = InMemoryVectorStore.from_documents(
        documents=doc_splits,
        embedding=OpenAIEmbeddings(),
    )
    return {"vectorstore": vectorstore}


def retrieve_and_generate(state):
    """
    Retrieve documents from vectorstore and generate answer
    """
    question = state["question"]
    vectorstore = state["vectorstore"]

    retriever = vectorstore.as_retriever()

    prompt = hub.pull("rlm/rag-prompt")
    llm = ChatOpenAI(model_name="gpt-3.5-turbo", temperature=0)

    # fetch relevant documents
    docs = retriever.invoke(question)  # format prompt
    formatted = prompt.invoke(
        {"context": docs, "question": question})  # generate answer
    answer = llm.invoke(formatted)
    return {"answer": answer}


# Graph
workflow = StateGraph(GraphState)

# Define the nodes
workflow.add_node("retrieve_and_generate", retrieve_and_generate)  # retrieve
workflow.add_node("scrape_blog_posts", scrape_blog_posts)  # scrape web
workflow.add_node("indexing", indexing)  # index

# Build graph
workflow.add_edge(START, "scrape_blog_posts")
workflow.add_edge("scrape_blog_posts", "indexing")
workflow.add_edge("indexing", "retrieve_and_generate")

workflow.add_edge("retrieve_and_generate", END)

# Compile
graph = workflow.compile()

File: ch10/js/rag-graph.js
import { Annotation, StateGraph } from '@langchain/langgraph';
import { CheerioWebBaseLoader } from '@langchain/community/document_loaders/web/cheerio';
import { RecursiveCharacterTextSplitter } from 'langchain/text_splitter';
import { MemoryVectorStore } from 'langchain/vectorstores/memory';
import { ChatOpenAI, OpenAIEmbeddings } from '@langchain/openai';
import * as hub from 'langchain/hub';
import { StringOutputParser } from '@langchain/core/output_parsers';

const GraphState = Annotation.Root({
  question: Annotation(),
  scrapedDocuments: Annotation(),
  vectorstore: Annotation(),
  answer: Annotation(),
});

const scrapeBlogPosts = async (state) => {
  const urls = [
    'https://blog.langchain.dev/top-5-langgraph-agents-in-production-2024/',
    'https://blog.langchain.dev/langchain-state-of-ai-2024/',
    'https://blog.langchain.dev/introducing-ambient-agents/',
  ];

  const loadDocs = async (urls) => {
    const docs = [];
    for (const url of urls) {
      const loader = new CheerioWebBaseLoader(url);
      const loadedDocs = await loader.load();
      docs.push(...loadedDocs);
    }
    return docs;
  };

  const scrapedDocuments = await loadDocs(urls);

  return { scrapedDocuments };
};

const indexing = async (state) => {
  const textSplitter = new RecursiveCharacterTextSplitter({
    chunkSize: 1000,
    chunkOverlap: 0,
  });

  const docSplits = await textSplitter.splitDocuments(state.scrapedDocuments);

  const vectorstore = new MemoryVectorStore(new OpenAIEmbeddings());

  await vectorstore.addDocuments(docSplits);

  console.log('vectorstore: ', vectorstore);

  return { vectorstore };
};

const retrieveAndGenerate = async (state) => {
  const { question, vectorstore } = state;

  const retriever = vectorstore.asRetriever();

  const prompt = await hub.pull('rlm/rag-prompt');

  const llm = new ChatOpenAI({ model: 'gpt-3.5-turbo', temperature: 0 });

  const docs = await retriever.invoke(question);

  const chain = prompt.pipe(llm).pipe(new StringOutputParser());

  const answer = await chain.invoke({ context: docs, question });

  console.log('answer: ', answer);

  return { answer };
};

const workflow = new StateGraph(GraphState)
  .addNode('retrieve_and_generate', retrieveAndGenerate)
  .addNode('scrape_blog_posts', scrapeBlogPosts)
  .addNode('indexing', indexing)
  .addEdge('__start__', 'scrape_blog_posts')
  .addEdge('scrape_blog_posts', 'indexing')
  .addEdge('indexing', 'retrieve_and_generate')
  .addEdge('retrieve_and_generate', '__end__');

const graph = workflow.compile();

await graph.invoke({ question: 'What are ambient agents?' });
